<a href="https://colab.research.google.com/github/treyra/CS155_miniproject/blob/master/Logistic_Regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# This notebook creates a simple logistic regression model on the data to experiment with different features and different normilizations

In [70]:
import numpy as np
from matplotlib import pyplot as plt
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import KFold
from sklearn.metrics import roc_auc_score
import pandas as pd

### Load the data:

In [2]:
#Use the imputed and normalized data (This link will need to be updated when we merge the branches)
dfImp = pd.read_csv('https://raw.githubusercontent.com/treyra/CS155_miniproject/data_visualization_v1/data_cleanup/traindata_imputed_normalized_minmax.csv',sep=',', header=0,index_col=0)

dfImp.head()

id  LATITUDE  LONGITUDE  STATE  DISCOVERY_TIME  ...  LABEL    0    1    2    3
0   0  0.988292  -0.985738      1        0.054167  ...      1  1.0  0.0  0.0  0.0
1   1 -0.401537   0.925104      0        0.464583  ...      4  0.0  0.0  0.0  1.0
2   2 -0.911691   1.066427      0        0.666667  ...      2  0.0  1.0  0.0  0.0
3   3 -0.814128   1.109280      0        0.506250  ...      4  0.0  0.0  0.0  1.0
4   4 -0.448755  -0.759942      1        0.549583  ...      2  0.0  1.0  0.0  0.0

[5 rows x 16 columns]

###Create the catgegory vs all data

In [4]:
#Make a copy for each label, set that labels value to 1, the rest to -1
#To change the features being passed in, make a dataframe with the new features, and replace dfImp with it.

oneData = dfImp.copy()
oneData.loc[(oneData['LABEL'] != 1,"LABEL")] = -1
#Make numpy arrays for regression
oneY = oneData['LABEL'].to_numpy()
oneData.drop('LABEL', axis=1)
oneX = oneData.to_numpy()

twoData = dfImp.copy()
twoData.loc[(twoData['LABEL'] != 2,"LABEL")] = -1
twoData.loc[(twoData['LABEL'] == 2,"LABEL")] = 1
#Make numpy arrays for regression
twoY = twoData['LABEL'].to_numpy()
twoData.drop('LABEL', axis=1)
twoX = twoData.to_numpy()

threeData = dfImp.copy()
threeData.loc[(threeData['LABEL'] != 3,"LABEL")] = -1
threeData.loc[(threeData['LABEL'] == 3,"LABEL")] = 1
threeY = threeData['LABEL'].to_numpy()
threeData.drop('LABEL', axis=1)
threeX = threeData.to_numpy()

fourData = dfImp.copy()
fourData.loc[(fourData['LABEL'] != 4,"LABEL")] = -1
fourData.loc[(fourData['LABEL'] == 4,"LABEL")] = 1
fourY = fourData['LABEL'].to_numpy()
fourData.drop('LABEL', axis=1)
fourX = fourData.to_numpy()

In [ ]:
len(oneX)

285382

###Logistic Regression and Cross Validation

In [5]:
#Regularization values to check via cross validation
#This can be swapped with some other parameter, the approach will be the came
numParams = 2
Cs = np.logspace(-3,1,numParams)

#Scores for in and out classification for each one vs all, determined using LogisticRegression.score()
scoreIn = np.zeros((4,numParams))
scoreOut = np.zeros((4,numParams))

#Run cross validation for each parameter!
num_folds = 10
for run in range(numParams):

    # Initialize kfold cross-validation object with 10 folds:
    kf = KFold(n_splits=num_folds)

    # Iterate through cross-validation folds for each model
    for train_index, test_index in kf.split(oneX):
        
        
        # Training and testing data points for this fold:
        x_train, x_test = oneX[train_index], oneX[test_index]
        y_train, y_test = oneY[train_index], oneY[test_index]

        # Instantiate a Ridge regression object (with little regularization, can change this)
        log = LogisticRegression(C = Cs[run])

        #Perform the fitting
        log.fit(x_train, y_train)

        #Evaluate errors 
        scoreIn[0,run] += log.score(x_train, y_train)
        scoreOut[0,run] += log.score(x_test, y_test)
    # Iterate through cross-validation folds for each model
    for train_index, test_index in kf.split(twoX):
        
        
        # Training and testing data points for this fold:
        x_train, x_test = twoX[train_index], twoX[test_index]
        y_train, y_test = twoY[train_index], twoY[test_index]

        # Instantiate a Ridge regression object (with little regularization, can change this)
        log = LogisticRegression(C = Cs[run])

        #Perform the fitting
        log.fit(x_train, y_train)

        #Evaluate errors 
        scoreIn[1,run] += log.score(x_train, y_train)
        scoreOut[1,run] += log.score(x_test, y_test)
    # Iterate through cross-validation folds for each model
    for train_index, test_index in kf.split(threeX):
        
        
        # Training and testing data points for this fold:
        x_train, x_test = threeX[train_index], threeX[test_index]
        y_train, y_test = threeY[train_index], threeY[test_index]

        # Instantiate a Ridge regression object (with little regularization, can change this)
        log = LogisticRegression(C = Cs[run])

        #Perform the fitting
        log.fit(x_train, y_train)

        #Evaluate errors 
        scoreIn[2,run] += log.score(x_train, y_train)
        scoreOut[2,run] += log.score(x_test, y_test)
    # Iterate through cross-validation folds for each model
    for train_index, test_index in kf.split(fourX):
        
        
        # Training and testing data points for this fold:
        x_train, x_test = fourX[train_index], fourX[test_index]
        y_train, y_test = fourY[train_index], fourY[test_index]

        # Instantiate a Ridge regression object (with little regularization, can change this)
        log = LogisticRegression(C = Cs[run])

        #Perform the fitting
        log.fit(x_train, y_train)

        #Evaluate errors 
        scoreIn[3,run] += log.score(x_train, y_train)
        scoreOut[3,run] += log.score(x_test, y_test)

scoreIn = scoreIn/num_folds
scoreOut = scoreOut/num_folds

print("Scores In")
print(scoreIn)
print("Scores Out")
print(scoreOut)



/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression


Scores In
[[0.91292606 0.91292606]
 [0.6141896  0.61418882]
 [0.85103826 0.85103826]
 [0.81656867 0.81656867]]
Scores Out
[[0.91309556 0.91309556]
 [0.61018603 0.61018603]
 [0.85103846 0.85103846]
 [0.81656852 0.81656852]]


Output from first run when Cs was accidentally set to .1-10**10

Scores In

[[0.91292294 0.91292294 0.91292294 0.91292294 0.91292294 0.91292294
  0.91292294 0.91292294 0.91292294 0.91292294 0.91292294 0.91292294
  0.91292294 0.91292294 0.91292294 0.91292294 0.91292294 0.91292294
  0.91292294 0.91292294]

 [0.6141896  0.6141896  0.6141896  0.6141896  0.6141896  0.6141896
  0.6141896  0.6141896  0.6141896  0.6141896  0.6141896  0.6141896
  0.6141896  0.6141896  0.6141896  0.6141896  0.6141896  0.6141896
  0.6141896  0.6141896 ]

 [0.85103826 0.85103826 0.85103826 0.85103826 0.85103826 0.85103826
  0.85103826 0.85103826 0.85103826 0.85103826 0.85103826 0.85103826
  0.85103826 0.85103826 0.85103826 0.85103826 0.85103826 0.85103826
  0.85103826 0.85103826]
 [0.81656867 0.81656867 0.81656867 0.81656867 0.81656867 0.81656867
  0.81656867 0.81656867 0.81656867 0.81656867 0.81656867 0.81656867
  0.81656867 0.81656867 0.81656867 0.81656867 0.81656867 0.81656867
  0.81656867 0.81656867]]

Scores Out

[[0.91292294 0.91292294 0.91292294 0.91292294 0.91292294 0.91292294
  0.91292294 0.91292294 0.91292294 0.91292294 0.91292294 0.91292294
  0.91292294 0.91292294 0.91292294 0.91292294 0.91292294 0.91292294
  0.91292294 0.91292294]

 [0.6141896  0.6141896  0.6141896  0.6141896  0.6141896  0.6141896
  0.6141896  0.6141896  0.6141896  0.6141896  0.6141896  0.6141896
  0.6141896  0.6141896  0.6141896  0.6141896  0.6141896  0.6141896
  0.6141896  0.6141896 ]

 [0.85103826 0.85103826 0.85103826 0.85103826 0.85103826 0.85103826
  0.85103826 0.85103826 0.85103826 0.85103826 0.85103826 0.85103826
  0.85103826 0.85103826 0.85103826 0.85103826 0.85103826 0.85103826
  0.85103826 0.85103826]

 [0.81656867 0.81656867 0.81656867 0.81656867 0.81656867 0.81656867
  0.81656867 0.81656867 0.81656867 0.81656867 0.81656867 0.81656867
  0.81656867 0.81656867 0.81656867 0.81656867 0.81656867 0.81656867
  0.81656867 0.81656867]]

In [44]:
#Make a copy for each label, set that labels value to 1, the rest to -1
#To change the features being passed in, make a dataframe with the new features, and replace dfImp with it.

featureDrop = dfImp.drop(['STATE','YEAR','FIPS_CODE','FIPS_NAME',
                  'SOURCE_REPORTING_UNIT_NAME','id','0','1','2','3'], axis=1)


oneData = featureDrop.copy()
oneData.loc[(oneData['LABEL'] != 1,"LABEL")] = -1
#Make numpy arrays for regression
oneY = oneData['LABEL'].to_numpy()

oneX = oneData.drop('LABEL', axis=1).to_numpy()
#Add weights to address there being less of each example
oneWeights = np.zeros(len(oneY))
numOnes = len(np.where(oneY==1)[0])
oneWeights[np.where(oneY==1)] = (len(oneY)-numOnes)/len(oneY)
oneWeights[np.where(oneY!=1)] = numOnes/len(oneY)

twoData = featureDrop.copy()
twoData.loc[(twoData['LABEL'] != 2,"LABEL")] = -1
twoData.loc[(twoData['LABEL'] == 2,"LABEL")] = 1
#Make numpy arrays for regression
twoY = twoData['LABEL'].to_numpy()

twoX = twoData.drop('LABEL', axis=1).to_numpy()
#Add weights to address there being less of each example
twoWeights = np.zeros(len(twoY))
numtwos = len(np.where(twoY==1)[0])
twoWeights[np.where(twoY==1)] = (len(twoY)-numtwos)/len(twoY)
twoWeights[np.where(twoY!=1)] = numtwos/len(twoY)

threeData = featureDrop.copy()
threeData.loc[(threeData['LABEL'] != 3,"LABEL")] = -1
threeData.loc[(threeData['LABEL'] == 3,"LABEL")] = 1
threeY = threeData['LABEL'].to_numpy()

threeX = threeData.drop('LABEL', axis=1).to_numpy()
#Add weights to address there being less of each example
threeWeights = np.zeros(len(threeY))
numthrees = len(np.where(threeY==1)[0])
threeWeights[np.where(threeY==1)] = (len(threeY)-numthrees)/len(threeY)
threeWeights[np.where(threeY!=1)] = numthrees/len(threeY)

fourData = featureDrop.copy()
fourData.loc[(fourData['LABEL'] != 4,"LABEL")] = -1
fourData.loc[(fourData['LABEL'] == 4,"LABEL")] = 1
fourY = fourData['LABEL'].to_numpy()

fourX = fourData.drop('LABEL', axis=1).to_numpy()
#Add weights to address there being less of each example
fourWeights = np.zeros(len(fourY))
numfours = len(np.where(fourY==1)[0])
fourWeights[np.where(fourY==1)] = (len(fourY)-numfours)/len(fourY)
fourWeights[np.where(fourY!=1)] = numfours/len(fourY)

In [46]:
#Regularization values to check via cross validation
#This can be swapped with some other parameter, the approach will be the came
numParams = 11
Cs = np.logspace(-10,1,numParams)

#Scores for in and out classification for each one vs all, determined using LogisticRegression.score()
scoreIn = np.zeros((4,numParams))
scoreOut = np.zeros((4,numParams))

#Run cross validation for each parameter!
num_folds = 10
for run in range(numParams):

    # Initialize kfold cross-validation object with 10 folds:
    kf = KFold(n_splits=num_folds)

    # Iterate through cross-validation folds for each model
    for train_index, test_index in kf.split(oneX):
        
        
        # Training and testing data points for this fold:
        x_train, x_test = oneX[train_index], oneX[test_index]
        y_train, y_test = oneY[train_index], oneY[test_index]
        weights = oneWeights[train_index]

        # Instantiate a Ridge regression object (with little regularization, can change this)
        log = LogisticRegression(C = Cs[run])

        #Perform the fitting
        log.fit(x_train, y_train,weights)

        #Evaluate errors 
        scoreIn[0,run] += log.score(x_train, y_train)
        scoreOut[0,run] += log.score(x_test, y_test)
    # Iterate through cross-validation folds for each model
    for train_index, test_index in kf.split(twoX):
        
        
        # Training and testing data points for this fold:
        x_train, x_test = twoX[train_index], twoX[test_index]
        y_train, y_test = twoY[train_index], twoY[test_index]
        weights = twoWeights[train_index]

        # Instantiate a Ridge regression object (with little regularization, can change this)
        log = LogisticRegression(C = Cs[run])

        #Perform the fitting
        log.fit(x_train, y_train,weights)

        #Evaluate errors 
        scoreIn[1,run] += log.score(x_train, y_train)
        scoreOut[1,run] += log.score(x_test, y_test)
    # Iterate through cross-validation folds for each model
    for train_index, test_index in kf.split(threeX):
        
        
        # Training and testing data points for this fold:
        x_train, x_test = threeX[train_index], threeX[test_index]
        y_train, y_test = threeY[train_index], threeY[test_index]
        weights = threeWeights[train_index]

        # Instantiate a Ridge regression object (with little regularization, can change this)
        log = LogisticRegression(C = Cs[run])

        #Perform the fitting
        log.fit(x_train, y_train,weights)

        #Evaluate errors 
        scoreIn[2,run] += log.score(x_train, y_train)
        scoreOut[2,run] += log.score(x_test, y_test)
    # Iterate through cross-validation folds for each model
    for train_index, test_index in kf.split(fourX):
        
        
        # Training and testing data points for this fold:
        x_train, x_test = fourX[train_index], fourX[test_index]
        y_train, y_test = fourY[train_index], fourY[test_index]
        weights = fourWeights[train_index]

        # Instantiate a Ridge regression object (with little regularization, can change this)
        log = LogisticRegression(C = Cs[run])

        #Perform the fitting
        log.fit(x_train, y_train,weights)

        #Evaluate errors 
        scoreIn[3,run] += log.score(x_train, y_train)
        scoreOut[3,run] += log.score(x_test, y_test)

scoreIn = scoreIn/num_folds
scoreOut = scoreOut/num_folds

print("Scores In")
print(scoreIn)
print("Scores Out")
print(scoreOut)

Scores In
[[0.58237841 0.58237841 0.58237841 0.56269175 0.65230429 0.65288397
  0.66918454 0.68907873 0.67906214 0.6769558  0.67675918]
 [0.47600372 0.47600372 0.47600372 0.51096742 0.62562694 0.62710761
  0.62691956 0.62714654 0.62716289 0.62717574 0.62717691]
 [0.43122935 0.43122935 0.43122935 0.40198124 0.56194138 0.54580137
  0.54578191 0.54574103 0.54557672 0.54553584 0.54553078]
 [0.37710142 0.35615991 0.3561603  0.36784522 0.64047719 0.60830201
  0.60778185 0.60852783 0.60968924 0.60982901 0.60984264]]
Scores Out
[[0.56512737 0.56512737 0.56512737 0.53952291 0.65353758 0.65259537
  0.66935889 0.68702992 0.6758834  0.67354617 0.6733955 ]
 [0.43247269 0.43247269 0.43247269 0.46540032 0.62015848 0.62587365
  0.62670061 0.62681975 0.62679873 0.62677069 0.6267742 ]
 [0.41685947 0.41685947 0.41685947 0.37485933 0.53625817 0.54579096
  0.54575592 0.54567883 0.54550012 0.54546859 0.54546508]
 [0.33981244 0.31949917 0.31949917 0.32998344 0.62132189 0.6079117
  0.60778205 0.60851791 0.609

Test accuracy from all features (no regularization)

[0.9794988397444669 0.7435116823214506 0.8662798073711897 0.8165686675754158]

In [47]:
#Best regularization for each parameter
print(np.max(scoreIn[0]),np.max(scoreIn[1]),np.max(scoreIn[2]),np.max(scoreIn[3]))
print(np.argmax(scoreIn,axis=1))

0.6890787304764945 0.6271769098652454 0.5619413762612553 0.6404771853699696
[ 7 10  4  4]


In [56]:
#Build model using the regularization (note we take for the 4th label c = 8 as regularization didn't have an effect so less is better)
Cs = np.logspace(-10,1,numParams)
model = [LogisticRegression(C = Cs[7]),LogisticRegression(C = Cs[10]),LogisticRegression(C = Cs[4]),LogisticRegression(C = Cs[4])]
model[0].fit(oneX, oneY,oneWeights)
print(1)
model[1].fit(twoX, twoY,twoWeights)
print(2)
model[2].fit(threeX, threeY,threeWeights)
print(3)
model[3].fit(fourX, fourY,fourWeights)
print(4)

#Print statements are so we can see where errors are in converging (if any).
#Seems to be on the lightning model, which has the best performance oddly


1
2
3
4


In [59]:
#Checking prediction works as we think
#Take random point
testPoints = oneX[0:20,:]
model[1].predict_proba(testPoints),model[1].predict(testPoints),twoY[0:20]

#testPositives = oneX[np.where(oneY == 1)]

#print(model[0].predict_proba(testPositives[0:50]))

(array([[0.58207108, 0.41792892],
        [0.35199495, 0.64800505],
        [0.3177252 , 0.6822748 ],
        [0.31928425, 0.68071575],
        [0.48945935, 0.51054065],
        [0.48370951, 0.51629049],
        [0.31021619, 0.68978381],
        [0.36150242, 0.63849758],
        [0.31980787, 0.68019213],
        [0.31888225, 0.68111775],
        [0.32316931, 0.67683069],
        [0.31772451, 0.68227549],
        [0.31517959, 0.68482041],
        [0.30856732, 0.69143268],
        [0.32386431, 0.67613569],
        [0.32421653, 0.67578347],
        [0.35796472, 0.64203528],
        [0.30869116, 0.69130884],
        [0.32323209, 0.67676791],
        [0.35454336, 0.64545664]]),
 array([-1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,
         1,  1,  1]),
 array([-1, -1,  1, -1,  1,  1, -1, -1, -1, -1,  1,  1, -1,  1, -1,  1, -1,
         1,  1, -1]))

In [72]:
#Testing model

#Predictions 
predictions = np.zeros((len(oneX),4))

#Predict the probabilities for each point
for i in range(4):
    predictions[:,i] = model[i].predict_proba(oneX)[:,1]

#Normalize each row
predictions = predictions / np.linalg.norm(predictions, axis=1,ord=1).reshape(len(oneY),1)

#Score
trainAUC = roc_auc_score(featureDrop['LABEL'].to_numpy(), predictions, multi_class='ovr')
trainAUC

0.6366257661720174

In [14]:
#Now running on the test data to predict
#Use the imputed and normalized data (This link will need to be updated when we merge the branches)
dfTest = pd.read_csv('https://raw.githubusercontent.com/treyra/CS155_miniproject/data_visualization_v1/data_cleanup/testdata_imputed_normalized_minmax.csv',sep=',', header=0,index_col=0)

dfTest.head()

id  LATITUDE  LONGITUDE  ...  SOURCE_REPORTING_UNIT_NAME      YEAR       DAY
0  285382 -0.232600  -0.828839  ...                   -0.303789  1.058824  0.002732
1  285383 -0.335939  -0.760377  ...                    0.320212  1.058824  0.002732
2  285384  0.945131  -0.982574  ...                   -0.871063  1.058824  0.002732
3  285385 -0.817080   0.998962  ...                    2.419125  1.058824  0.005464
4  285386 -0.682598  -0.695973  ...                   -0.681972  1.058824  0.005464

[5 rows x 11 columns]

In [80]:
testDrop = dfTest.drop(['STATE','YEAR','FIPS_CODE','FIPS_NAME',
                  'SOURCE_REPORTING_UNIT_NAME','id'], axis=1)
testX = testDrop.to_numpy()

In [84]:
#Predictions on test data

#Predictions 

predictionsTest = np.zeros((len(testX),4))

#Predict the probabilities for each point
for i in range(4):
    predictionsTest[:,i] = model[i].predict_proba(testX)[:,1]

#Normalize each row
predictionsTest = predictionsTest / np.linalg.norm(predictionsTest, axis=1,ord=1).reshape(len(testX),1)

#Add id
ids = np.zeros((len(testX),1))
ids[:,0] = dfTest["id"].to_numpy()
predictionsTest = np.hstack((ids,predictionsTest))
#Save to a csv
predictionsTest.tofile('WILDFIRES_TEST.csv', sep = ',')

In [86]:
from google.colab import drive
drive.mount('drive')
#Save to a csv

Drive already mounted at drive; to attempt to forcibly remount, call drive.mount("drive", force_remount=True).


In [89]:
df = pd.DataFrame(predictionsTest)
df.to_csv('WILDFIRES_TEST.csv', sep = ',')
!cp WILDFIRES_TEST.csv "drive/My Drive/"

In [90]:
df

0         1         2         3         4
0      285382.0  0.146393  0.282179  0.284065  0.287363
1      285383.0  0.138295  0.289212  0.285035  0.287458
2      285384.0  0.239881  0.224618  0.263572  0.271930
3      285385.0  0.118297  0.357493  0.267693  0.256518
4      285386.0  0.102760  0.309488  0.293493  0.294259
...         ...       ...       ...       ...       ...
73030  358412.0  0.190476  0.305899  0.257590  0.246034
73031  358413.0  0.219848  0.232088  0.273195  0.274869
73032  358414.0  0.281112  0.194960  0.259296  0.264633
73033  358415.0  0.316674  0.178080  0.249360  0.255886
73034  358416.0  0.369093  0.148829  0.236360  0.245717

[73035 rows x 5 columns]